torchtune/models/llama3/_tokenizer.py

```diff
@@ -0,0 +1,181 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Dict, List, Optional, Tuple
+
+from torchtune.data import Message, truncate
+from torchtune.modules.tokenizers import (
+    ModelTokenizer,
+    parse_hf_tokenizer_json,
+    TikTokenBaseTokenizer,
+)
+
+
+CL100K_PATTERN = r"""(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+"""  # noqa
+
+PAD_ID = 0
+
+
+class Llama3Tokenizer(ModelTokenizer):
+    """
+    tiktoken tokenizer configured with Llama3 Instruct's special tokens, as described in
+    https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-3
+
+    Args:
+        path (str): Path to pretrained tiktoken tokenizer file.
+        special_tokens_path (str): Path to ``tokenizer.json`` from Hugging Face
+            model files that contains all registered special tokens.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        special_tokens_path: str,
```

Comment:
So I understand the purpose of this and it is nice that we can just directly load from HF. But I also don't love that (a) we are very tied to their specific tokenizer representation (and not just generic json) and (b) it kinda abstracts away the actual special tokens a bit (though I guess you do define them below). Another approach here would be to pass `special_tokens: Dict[str, int]` and leave it to the builder to take the tokenizer json path and unpack it. Tbh I don't have a strong opinion here though

Reply:
Ohh the second option sounds good. We will provide the function to load and build the special tokens via the HF file but allow people to pass in a dict if they want. Smart cookie.